From 456b52363338f26c364159dcc8138a1dedff09fd Mon Sep 17 00:00:00 2001
From: Aymen Zayet <aymen.zayet@intel.com>
Date: Fri, 17 Aug 2012 11:54:35 +0200
Subject: [PATCH 19/78] wl12xx: release wake lock when removing the interface.

BZ: 48052

When the hard IRQ is called, a wake lock is held and the thread irq
is scheduled later time. This one is supposed to release the wakelock.
If a recovery happens in between, the interrupt are disabled without
irq_synchronize(). As a consequence, the thread irq will be cancelled
and the wake lock remains held.

Moreover, recovery calls __wl1271_op_remove_interface() which clears
wl->flags, making the call to wake_lock() / wake_unlock() unbalanced.

To fix this issue, WL1271_FLAG_WAKE_LOCK flag should be checked on
wl1271_op_remove_interface() and the wake_lock should be called
accordingly.

Change-Id: Ie7657ec3ba67013b27e1df69d25ee35e86762c87
Orig-Change-Id: I520daf004339ca7339b31de60be066981ba59a5e
Signed-off-by: Aymen Zayet <aymen.zayet@intel.com>
Signed-off-by: Axel Haslam <axelx.haslam@intel.com>
Reviewed-on: http://android.intel.com:8080/62660
Reviewed-by: Regairaz, Julie <julie.regairaz@intel.com>
Tested-by: Champciaux, Nicolas <nicolas.champciaux@intel.com>
Reviewed-by: buildbot <buildbot@intel.com>
Tested-by: buildbot <buildbot@intel.com>
Reviewed-on: http://android.intel.com:8080/77976
Reviewed-by: Li, XiaojingX <xiaojingx.li@intel.com>
Tested-by: Li, XiaojingX <xiaojingx.li@intel.com>
---
 wl12xx-compat/drivers/net/wireless/wl12xx/main.c |    6 ++++++
 1 files changed, 6 insertions(+), 0 deletions(-)

diff --git a/wl12xx-compat/drivers/net/wireless/wl12xx/main.c b/wl12xx-compat/drivers/net/wireless/wl12xx/main.c
index 2afb0a8..3e30f12 100644
--- a/wl12xx-compat/drivers/net/wireless/wl12xx/main.c
+++ b/wl12xx-compat/drivers/net/wireless/wl12xx/main.c
@@ -2730,6 +2730,12 @@ deinit:
 	cancel_work_sync(&wlvif->rx_streaming_disable_work);
 
 	mutex_lock(&wl->mutex);
+
+#ifdef CONFIG_HAS_WAKELOCK
+	if (test_and_clear_bit(WL1271_FLAG_WAKE_LOCK, &wl->flags))
+			wake_unlock(&wl->wake_lock);
+#endif
+
 }
 
 static void wl1271_op_remove_interface(struct ieee80211_hw *hw,
-- 
1.7.9

